### PR TITLE
Add new durability modes for improving performance + other goodies

### PR DESCRIPTION
**Pull request by svanoort**
Replaces #170 

# Functional tasks
- [x] Pass and use DurabilityHints to set up storage
- [x] Use durability hint to decide when we should persist program and what storage to use for node (plus when to persist them).
- [x] Correctly defer writing FlowNode actions, with test cases
- [x] Provide testcases for the key durability modes covering: resume build, complete build, try to resume build when we can't do so due to system failure.
- [x] Tests verify no dangling tasks left
- [x] Test that hard failure of the master (not clean shutdown) doesn't hose things upon restart (clean failures)
- [x] CPS Program persists before restart if we're set up to survive clean restarts but not continuously persistent
- [x] Tests pass for dirty restarts
- [x] Optional: try to cleanly terminate flow graphs with incomplete persistence state (best-effort basis)
- [x] Apply any refactors to the DurabilityHint API

# Administrivia
- [x] Re-merge master
- [x] Deploy the upstream SNAPSHOTs

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java [hunk 1]
```diff
@@ -361,15 +381,21 @@ public CpsFlowExecution(String script, FlowExecutionOwner owner) throws IOExcept
...
         this.script = script;
         this.sandbox = sandbox;
+        this.durabilityHint = durabilityHint;
         this.storage = createStorage();
+        this.storage.setAvoidAtomicWrite(!this.getDurabilityHint().isAtomicWrite());
         Authentication auth = Jenkins.getAuthentication();
         this.user = auth.equals(ACL.SYSTEM) ? null : auth.getName();
     }
 
+    public CpsFlowExecution(String script, boolean sandbox, FlowExecutionOwner owner) throws IOException {
```
Comment (abayer) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java]: Did you want a `@Deprecated` on here?

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowDefinition.java [hunk 2]
```diff
@@ -75,6 +77,11 @@ public CpsFlowDefinition(String script, boolean sandbox) {
         this.sandbox = sandbox;
     }
 
+    @DataBoundSetter
+    public void setDurabilityHint(FlowDurabilityHint durabilityHint) {
+        super.setDurabilityHint(durabilityHint);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowDefinition.java]: This should be setting a `private` field.

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinition.java [hunk 3]
```diff
@@ -68,6 +69,11 @@
         this.scriptPath = scriptPath;
     }
 
+    @DataBoundSetter
+    public void setDurabilityHint(FlowDurabilityHint durabilityHint) {
+        super.setDurabilityHint(durabilityHint);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinition.java]: ditto

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsBodyExecution.java [hunk 4]
```diff
@@ -119,6 +119,8 @@
                 sn.addAction(a);
         }
 
+        CpsFlowExecution.maybeAutoPersistNode(sn);  // Hacky but helpful
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsBodyExecution.java]: But not the most helpful comment. :-) Is there something a reader should know? I for one have no idea what this is about.

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowDefinition.java [hunk 5]
```diff
@@ -106,7 +113,7 @@ public CpsFlowExecution create(FlowExecutionOwner owner, TaskListener listener,
                 return ((CpsFlowFactoryAction2) a).create(this, owner, actions);
             }
         }
-        return new CpsFlowExecution(sandbox ? script : ScriptApproval.get().using(script, GroovyLanguage.get()), sandbox, owner);
+        return new CpsFlowExecution(sandbox ? script : ScriptApproval.get().using(script, GroovyLanguage.get()), sandbox, owner, getDurabilityHint());
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowDefinition.java]: …and this reading it.

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java [hunk 6]
```diff
@@ -58,6 +58,8 @@
 import org.jboss.marshalling.Unmarshaller;
 import org.jenkinsci.plugins.workflow.actions.ErrorAction;
 import org.jenkinsci.plugins.workflow.cps.persistence.PersistIn;
+import org.jenkinsci.plugins.workflow.flow.BlockableResume;
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java]: Sorry, where is this defined? I do not see it in https://github.com/jenkinsci/workflow-api-plugin/pull/55. Did you forget to `git add`?

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java (hunk 1 above)
Comment (jglick) [reply src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java]: (Maybe time for a builder?)

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java [hunk 7]
```diff
@@ -559,34 +594,93 @@ int approximateNodeCount() {
...
+            this.heads.clear();
+            this.startNodes.clear();
+            FlowHead head = new FlowHead(this);
+            heads.put(head.getId(), head);
+            try {
+                FlowStartNode start = new FlowStartNode(this, iotaStr());
+                startNodes.push(start);
+                head.newStartNode(start);
+            } catch (IOException e) {
+                LOGGER.log(Level.FINE, "Failed to persist", e);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java]: Not a `WARNING`?

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java [hunk 8]
```diff
@@ -559,34 +594,93 @@ int approximateNodeCount() {
...
             headsSerial = null;
             // Same for startNodes:
             startNodes = new Stack<BlockStartNode>();
             for (String id : startNodesSerial) {
-                startNodes.add((BlockStartNode) storage.getNode(id));
+                FlowNode node = storage.getNode(id);
+                if (node != null) {
+                    startNodes.add((BlockStartNode) storage.getNode(id));
+                } else {
+                    // TODO if possible, consider tyring to close out unterminated blocks to keep existing graph history
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java]: typo

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java [hunk 9]
```diff
@@ -559,34 +594,93 @@ int approximateNodeCount() {
...
             initializeStorage();
             try {
                 if (!isComplete()) {
-                    loadProgramAsync(getProgramDataFile());
+                    if (canResume()) {
+                        loadProgramAsync(getProgramDataFile());
+                    } else {
+                        // TODO if possible, consider tyring to close out unterminated blocks to keep existing graph history
+                        // That way we can visualize the graph in some error cases.
+                        LOGGER.log(Level.WARNING, "Pipeline state not properly persisted, cannot resume "+owner.getUrl());
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java]: Not sure we really need a warning in the system log for this, so long as a clear error is printed to the build log. This is a pretty normal condition after all.

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java [hunk 10]
```diff
@@ -1379,71 +1509,82 @@ public void marshal(Object source, HierarchicalStreamWriter w, MarshallingContex
...
-                    } else
-                    if (nodeName.equals("loadedScripts")) {
-                        Map loadedScripts = readChild(reader, context, Map.class, result);
-                        setField(result, "loadedScripts", loadedScripts);
-                    } else
-                    if (nodeName.equals("timings")) {
-                        Map timings = readChild(reader, context, Map.class, result);
-                        setField(result, "timings", timings);
-                    } else
-                    if (nodeName.equals("sandbox")) {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java]: Hard to follow diff here, could original indentation be restored?

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java [hunk 11]
```diff
@@ -1536,4 +1710,108 @@ private void setField(CpsFlowExecution result, String fieldName, Object value) {
...
+                            FlowNode n = context.get(FlowNode.class);
+                            if (n != null) {
+                                owner.getListener().getLogger().println("Terminating " + n.getDisplayFunctionName());
+                            }
+                        } catch (Exception x) {
+                            LOGGER.log(Level.FINE, null, x);
+                        }
+                    }
+                }
+                @Override public void onFailure(Throwable t) {}
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java]: Need to handle.

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java [hunk 12]
```diff
@@ -1536,4 +1710,108 @@ private void setField(CpsFlowExecution result, String fieldName, Object value) {
...
+            final Throwable x = new FlowInterruptedException(Result.ABORTED);
+            Futures.addCallback(this.getCurrentExecutions(/* cf. JENKINS-26148 */true), new FutureCallback<List<StepExecution>>() {
+                @Override public void onSuccess(List<StepExecution> l) {
+                    for (StepExecution e : Iterators.reverse(l)) {
+                        StepContext context = e.getContext();
+                        context.onFailure(x);
+                        try {
+                            FlowNode n = context.get(FlowNode.class);
+                            if (n != null) {
+                                owner.getListener().getLogger().println("Terminating " + n.getDisplayFunctionName());
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java]: I guess this is where you would want to `StepExecution.stop` or something?

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsStepContext.java [hunk 13]
```diff
@@ -522,7 +522,13 @@ synchronized boolean switchToAsyncMode() {
     @Override public ListenableFuture<Void> saveState() {
         try {
             final SettableFuture<Void> f = SettableFuture.create();
-            getFlowExecution().runInCpsVmThread(new FutureCallback<CpsThreadGroup>() {
+            CpsFlowExecution exec = getFlowExecution();
+            if (!exec.getDurabilityHint().isPersistWithEveryStep()) {
+                f.set(null);
+                return f;
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsStepContext.java]: Can use `ImmediateFuture` or something like that IIRC.

File: src/main/java/org/jenkinsci/plugins/workflow/cps/replay/ReplayFlowFactoryAction.java [hunk 14]
```diff
@@ -59,7 +59,7 @@
     @Override public CpsFlowExecution create(FlowDefinition def, FlowExecutionOwner owner, List<? extends Action> actions) throws IOException {
         String script = replacementMainScript;
         replacementMainScript = null; // minimize build.xml size
-        return new CpsFlowExecution(script, sandbox, owner);
+        return new CpsFlowExecution(script, sandbox, owner, def.getDurabilityHint());
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/replay/ReplayFlowFactoryAction.java]: Should look up the durability of the original build, _not_ the current setting.

File: src/test/java/org/jenkinsci/plugins/workflow/cps/FlowDurabilityTest.java [hunk 15]
```diff
@@ -0,0 +1,672 @@
...
+        }
+    }
+
+    /** Sanity check that fully durable pipelines shutdown and restart cleanly */
+    @Test
+    public void testFullyDurableSurvivesCleanRestart() throws Exception {
+        final String jobName = "survivesEverything";
+        final String[] logStart = new String[1];
+
+        story.addStep(new Statement() {
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/FlowDurabilityTest.java]: BTW `.then` is more readable.

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java [hunk 16]
```diff
@@ -324,6 +333,18 @@
     /** Actions to add to the {@link FlowStartNode}. */
     transient final List<Action> flowStartNodeActions = new ArrayList<Action>();
 
+    /** If true, pipeline is forbidden to resume even if it can. */
+    public boolean isResumeBlocked() {
+        return resumeBlocked;
+    }
+
+    public void setResumeBlocked(boolean resumeBlocked) {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java]: Who calls this? It seems only test code for now? What is the expected UX?

Comment (svanoort) [reply src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java]: Downstream - used in WorkflowJob

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java (hunk 6 above)
Comment (svanoort) [reply src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java]: Committeed but hadn't been pushed out in the upstream -- you caught me in the midst of pushing out something small, but it's there now.

**Comment by jglick** [comment]
needs merge conflict resolution BTW

**Comment by svanoort** [comment]
@jglick Fixed merge conflicts, I was waiting on another piece in order to resolve the latest release's introduced issues.

File: pom.xml [hunk 17]
```diff
@@ -66,10 +66,11 @@
         <java.level>8</java.level>
         <no-test-jar>false</no-test-jar>
         <git-plugin.version>3.1.0</git-plugin.version>
-        <workflow-support-plugin.version>2.17-durability-v3-20171212.214810-1</workflow-support-plugin.version>
+        <workflow-support-plugin.version>2.17-durability-beta-1</workflow-support-plugin.version>
         <scm-api-plugin.version>2.0.8</scm-api-plugin.version>
         <groovy-cps.version>1.21</groovy-cps.version>
-        <jenkins-test-harness.version>2.33-20171212.224055-2</jenkins-test-harness.version>
+        <!-- Just until https://github.com/jenkinsci/jenkins-test-harness/pull/91 is released -->
```
Comment (jglick) [inline pom.xml]: obsolete comment

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowDefinition.java [hunk 18]
```diff
@@ -113,7 +112,9 @@ public CpsFlowExecution create(FlowExecutionOwner owner, TaskListener listener,
                 return ((CpsFlowFactoryAction2) a).create(this, owner, actions);
             }
         }
-        return new CpsFlowExecution(sandbox ? script : ScriptApproval.get().using(script, GroovyLanguage.get()), sandbox, owner, getDurabilityHint());
+        Queue.Executable exec = owner.getExecutable();
+        FlowDurabilityHint hint = (exec instanceof Run) ? DurabilityHintProvider.suggestedFor(((Run)exec).getParent()) : GlobalDefaultFlowDurabilityLevel.getDefaultDurabilityHint();
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowDefinition.java]: `DurabilityHintProvider` should be operating on `FlowExecutionOwner`, not `Run`. As written, it will be unusable for (say) a `/script` console-like Pipeline runner.

File: src/test/java/org/jenkinsci/plugins/workflow/cps/FlowDurabilityTest.java [hunk 19]
```diff
@@ -135,7 +138,8 @@ static WorkflowRun createAndRunBasicJob(Jenkins jenkins, String jobName, FlowDur
                 "semaphore 'halt' \n" +
                 "} \n" +
                 "echo 'I like chese'\n", false);
-        def.setDurabilityHint(durabilityHint);
+        TestDurabilityHintProvider provider = Jenkins.getInstance().getExtensionList(TestDurabilityHintProvider.class).get(0);
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/FlowDurabilityTest.java]: BTW `ExtensionList.lookup`; and a `TODO` comment to use `lookupSingleton` when available (later than your current baseline I think).

File: src/test/java/org/jenkinsci/plugins/workflow/cps/FlowDurabilityTest.java [hunk 20]
```diff
@@ -648,23 +654,76 @@ public void evaluate() throws Throwable {
...
     }
 
+    /** Test interrupting build by randomly dying at unpredictable times. */
     @Test
-    @Ignore
-    @TimedRepeatRule.RepeatForTime(repeatMillis = 300_000)
-    public void fuzzTesting() throws Exception {
-
-        long startTime = System.nanoTime();
+    @Ignore //Too long to run as part of main suite
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/FlowDurabilityTest.java]: you can put comments in the `Ignore.value`

File: src/test/java/org/jenkinsci/plugins/workflow/cps/FlowDurabilityTest.java [hunk 21]
```diff
@@ -648,23 +654,76 @@ public void evaluate() throws Throwable {
...
+                int time = new Random().nextInt((int) millisDuration);
+                WorkflowRun run = job.scheduleBuild2(0).getStartCondition().get();
+                Thread.sleep(time);
+                logStart[0] = JenkinsRule.getLog(run);
+                nodesOut.clear();
+                nodesOut.addAll(new DepthFirstScanner().allNodes(run.getExecution()));
+                nodesOut.sort(FlowScanningUtils.ID_ORDER_COMPARATOR);
+            }
+            });
+        story.addStep(new Statement() {
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/FlowDurabilityTest.java]: or `.then`

File: src/test/java/org/jenkinsci/plugins/workflow/cps/FlowDurabilityTest.java [hunk 22]
```diff
@@ -648,23 +654,76 @@ public void evaluate() throws Throwable {
...
+                if (run.isBuilding()) {
+                    story.j.waitForCompletion(run);
+                    Assert.assertEquals(Result.SUCCESS, run.getResult());
+                } else {
+                    verifyCompletedCleanly(story.j.jenkins, run);
                 }
+                assertIncludesNodes(nodesOut, run);
+                story.j.assertLogContains(logStart[0], run);
+                try {
+                    FlowExecutionList.get().unregister(run.asFlowExecutionOwner());
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/FlowDurabilityTest.java]: Why is this necessary? If it completed, it should be gone from FEL right? And why would you care at the end of a test anyway?

File: src/test/java/org/jenkinsci/plugins/workflow/cps/FlowDurabilityTest.java [hunk 23]
```diff
@@ -648,23 +654,76 @@ public void evaluate() throws Throwable {
...
+                } catch (Exception ex){}
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/FlowDurabilityTest.java]: Let any exception thrown here be thrown up and fail the test.

File: src/test/java/org/jenkinsci/plugins/workflow/cps/replay/ReplayActionTest.java [hunk 24]
```diff
@@ -58,6 +58,7 @@
 import org.jenkinsci.plugins.workflow.test.steps.SemaphoreStep;
 import static org.junit.Assert.*;
 import org.junit.ClassRule;
+import org.junit.Ignore;
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/replay/ReplayActionTest.java]: ?

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowDefinition.java (hunk 18 above)
Comment (svanoort) [reply src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowDefinition.java]: AFAIK no such thing exists, nor have I heard of plans to provide one?

**Comment by svanoort** [comment]
Looking into the test failure (some weirdness with the memory test that is difficult to diagnose)

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowDefinition.java (hunk 18 above)
Comment (jglick) [reply src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowDefinition.java]: Does not exist currently, but all the APIs are designed with the intent of making it possible to provide such an engine.

File: src/test/java/org/jenkinsci/plugins/workflow/cps/FlowDurabilityTest.java [hunk 25]
```diff
@@ -500,7 +536,38 @@ public void evaluate() throws Throwable {
...
+            @Override
+            public void evaluate() throws Throwable {
+                WorkflowRun run = createAndRunSleeperJob(story.j.jenkins, "durableAgainstClean", FlowDurabilityHint.PERFORMANCE_OPTIMIZED);
+                Assert.assertEquals(FlowDurabilityHint.PERFORMANCE_OPTIMIZED, run.getExecution().getDurabilityHint());
+                logStart[0] = JenkinsRule.getLog(run);
+                CpsFlowExecution exec = (CpsFlowExecution)(run.getExecution());
+
+                // Ensure the storage file is unreadable
+                try (FileChannel fis = new FileOutputStream(new File(exec.getStorageDir(), "flowNodeStore.xml")).getChannel()) {
+                    fis.truncate(5); // Leave a tiny bit just to make things more interesting
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/FlowDurabilityTest.java]: heh

**Comment by jglick** [comment]
> weirdness with the memory test that is difficult to diagnose

BTW if the test fails reproducibly in this branch, yet you are not getting root reference chains printed with the explanation, you can try

* bisecting until you find which change introduced the issue (may not be feasible due to use of multiple `beta` versions rather than a consistent set of `SNAPSHOT` versions)
* `jmap -histo:live` the `surefirebooter` process while it is struggling to collect, and using the heap profiler of your choice to dig in

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java [hunk 26]
```diff
@@ -272,6 +272,14 @@
     /** Subdirectory string where we store {@link FlowNode}s */
     private String storageDir = null;
 
+    static {
+        // Prevents the sole Timeout thread from being created where it will get a Groovy classloader and cause memory leaks
+        try (Timeout tim = Timeout.limit(1, TimeUnit.SECONDS)){
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java]: This is better fixed in `Timeout`.

Comment (jglick) [reply src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java]: cf. https://github.com/jenkinsci/workflow-support-plugin/pull/53

Comment (svanoort) [reply src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java]: It's coming, just takes a little bit for the deployment chain to catch up.

**Comment by svanoort** [comment]
Requesting final review -- all tests are now green.

File: src/main/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImpl.java [hunk 27]
```diff
@@ -274,7 +273,7 @@ Object sanitizeObjectAndRecordMutation(@CheckForNull Object o, @CheckForNull Env
     @Nonnull
     Map<String,Object> sanitizeMapAndRecordMutation(@Nonnull Map<String, Object> mapContents, @CheckForNull EnvVars variables) {
         // Package scoped so we can test it directly
-        HashMap<String, Object> output = new HashMap<String, Object>();
+        HashMap<String, Object> output = Maps.newHashMapWithExpectedSize(mapContents.size());
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImpl.java]: How is this related to the rest of the PR? Seems like something that should have been filed separately.

File: src/main/java/org/jenkinsci/plugins/workflow/cps/DSL.java [hunk 28]
```diff
@@ -423,7 +439,7 @@ private static Object flattenGString(Object v) {
             return mutated ? r : v;
         } else if (v instanceof Map) {
             boolean mutated = false;
-            Map<Object,Object> r = new LinkedHashMap<>();
+            Map<Object,Object> r = new LinkedHashMap<>(preallocatedHashmapCapacity(((Map) v).size()));
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/DSL.java]: ditto

File: src/main/java/org/jenkinsci/plugins/workflow/cps/DSL.java [hunk 29]
```diff
@@ -383,16 +388,27 @@ private void reportMissingContextVariableException(CpsStepContext context, Missi
...
         final Closure body;
 
         private NamedArgsAndClosure(Map<?,?> namedArgs, Closure body) {
-            this.namedArgs = new LinkedHashMap<String,Object>();
+            this.namedArgs = new LinkedHashMap<String,Object>(preallocatedHashmapCapacity(namedArgs.size()));
             this.body = body;
 
             for (Map.Entry<?,?> entry : namedArgs.entrySet()) {
-                String k = entry.getKey().toString(); // coerces GString and more
+                String k = entry.getKey().toString().intern(); // coerces GString and more
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/DSL.java]: ditto

File: src/main/java/org/jenkinsci/plugins/workflow/cps/DSL.java [hunk 30]
```diff
@@ -383,16 +388,27 @@ private void reportMissingContextVariableException(CpsStepContext context, Missi
...
+        }
+    }
+
     static class NamedArgsAndClosure {
         final Map<String,Object> namedArgs;
         final Closure body;
 
         private NamedArgsAndClosure(Map<?,?> namedArgs, Closure body) {
-            this.namedArgs = new LinkedHashMap<String,Object>();
+            this.namedArgs = new LinkedHashMap<String,Object>(preallocatedHashmapCapacity(namedArgs.size()));
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/DSL.java]: ditto

File: src/main/resources/org/jenkinsci/plugins/workflow/cps/CpsFlowDefinition/config.jelly [hunk 31]
```diff
@@ -33,4 +33,5 @@
     <f:block>
         <a href="pipeline-syntax" target="_blank">${%Pipeline Syntax}</a>
     </f:block>
+
```
Comment (jglick) [inline src/main/resources/org/jenkinsci/plugins/workflow/cps/CpsFlowDefinition/config.jelly]: revert

File: src/main/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImpl.java (hunk 27 above)
Comment (svanoort) [reply src/main/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImpl.java]: @jglick It's due to the BulkFlowNodeStorage we now use as of this PR holding more nodes in memory (potentially).  This reduces the memory impact of that.

File: src/main/java/org/jenkinsci/plugins/workflow/cps/DSL.java [hunk 32]
```diff
@@ -64,6 +64,8 @@
 import org.jenkinsci.plugins.structs.describable.DescribableParameter;
 import org.jenkinsci.plugins.structs.describable.UninstantiatedDescribable;
 import static org.jenkinsci.plugins.workflow.cps.ThreadTaskResult.*;
+
+import org.jenkinsci.plugins.workflow.actions.TimingAction;
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/DSL.java]: Unused import?

File: src/main/java/org/jenkinsci/plugins/workflow/cps/FlowHead.java [hunk 33]
```diff
@@ -106,14 +111,17 @@ void newStartNode(FlowStartNode n) throws IOException {
...
+        execution.storage.storeNode(head, false);
     }
 
     void setNewHead(FlowNode v) {
         try {
+            if (this.head != null) {
+                CpsFlowExecution.maybeAutoPersistNode(head);
+            }
+            execution.storage.storeNode(v, true);
+            v.addAction(new TimingAction());
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/FlowHead.java]: [JENKINS-38737](https://issues.jenkins-ci.org/browse/JENKINS-38737) FTR
